### PR TITLE
Implement Azure Storage Blob Connector

### PR DESCRIPTION
**Pull request by LakshanSS**
Fix https://github.com/wso2-enterprise/choreo/issues/2741

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: modules/blobs/data_mappings.bal [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+# + response - validated http response
+# + return - Returns AccountInformation type
+isolated function convertResponseToAccountInformationType(http:Response response) 
+        returns @tainted AccountInformationResult|error {
+    AccountInformationResult accountInformation = {};
+    accountInformation.accountKind = response.getHeader(X_MS_ACCOUNT_KIND);
+    accountInformation.skuName = response.getHeader(X_MS_SKU_NAME);
+    accountInformation.isHNSEnabled = response.getHeader(X_MS_IS_HNS_ENABLED);
+    accountInformation.responseHeaders = getHeaderMapFromResponse(<http:Response>response);
+    return accountInformation;
```
Comment (SanduDS) [inline modules/blobs/data_mappings.bal:60]: AccountInformationResult.skuName is string. But sometimes, get header method can send an error like HeaderNotFoundError. then will it be set to the AccountInformationResult.skuName?

File: modules/blobs/data_mappings.bal [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+# Creates ContainerACLResult from http response.
+# 
+# + response - validated http response
+# + return - Returns ContainerACLResult type
+isolated function convertResponseToContainerACLResult(http:Response response) 
+        returns @tainted ContainerACLResult|error {                    
+    ContainerACLResult containerACLResult = {};                  
+    containerACLResult.eTag = response.getHeader(ETAG);
+    containerACLResult.lastModified = response.getHeader(LAST_MODIFIED);
+
```
Comment (SanduDS) [inline modules/blobs/data_mappings.bal:113]: Can remove spaces?

File: modules/blobs/data_mappings.bal [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+# + blobListJson - list of blobs in json format
+# + return - Returns Blob array
+isolated function convertJSONToBlobArray(json|error blobListJson) returns Blob[]|error {
+    Blob[] blobList = [];
+    if (blobListJson is json[]) { // When there are multiple blobs, it will be a json[]
+        foreach json blobJsonObject in blobListJson {
+            Blob blob = check convertJSONToBlobType(blobJsonObject);
+            arrlib:push(blobList, blob);
+        }
+    } else if (blobListJson is json){ // When there is only one blob, it will be a json
```
Comment (SanduDS) [inline modules/blobs/data_mappings.bal:215]: check formatting

File: modules/blobs/endpoint.bal [hunk 4]
```diff
@@ -0,0 +1,939 @@
...
+        blobResult.blobContent = <byte[]>check handleGetBlobResponse(response);
+        blobResult.responseHeaders = getHeaderMapFromResponse(<http:Response>response);
+        return blobResult;
+    }
+
+    # Get Account Information of the azure storage account
+    # 
+    # + clientRequestId - Optional. Client request Id
+    # + return - If successful, returns AccountInformation. Else returns Error. 
+    remote function getAccountInformation(string? clientRequestId=()) returns @tainted AccountInformationResult|error {
```
Comment (SanduDS) [inline modules/blobs/endpoint.bal:154]: check formatting

File: modules/blobs/endpoint.bal [hunk 5]
```diff
@@ -0,0 +1,939 @@
...
+    # + containerName - name of the container
+    # + pageBlobName - name of the page blob
+    # + operation - It can be update or clear
+    # + range - Specifies the range of bytes to be written as a page. 
+    # + content - blob content
+    # + timeout - Optional. Timout value expressed in seconds
+    # + clientRequestId - Optional. Client generated value for correlating client side activities with requests received
+    #  by the server.
+    # + return - If successful, returns Response Headers. Else returns Error.
+    remote function putPage(string containerName, string pageBlobName, string operation, string range,
```
Comment (SanduDS) [inline modules/blobs/endpoint.bal:859]: check formatting

File: main.bal [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (RolandHewage) [inline main.bal:1]: Remove Unwanted files

File: modules/blobs/endpoint.bal [hunk 7]
```diff
@@ -0,0 +1,939 @@
...
+        http:Request request =  createRequest(optionsHolder.optionalHeaders);
+        map<string> uriParameterMap = optionsHolder.optionalURIParameters;
+        uriParameterMap[COMP] = LIST;
+
+        check prepareAuthorizationHeader(request, GET, self.authorizationMethod, self.accountName, self.accessKey, 
+                                            EMPTY_STRING, uriParameterMap);
+        string resourcePath = FORWARD_SLASH_SYMBOL;
+        string path = preparePath(self.authorizationMethod, self.sharedAccessSignature, uriParameterMap, resourcePath);
+        http:Response response = <http:Response> check self.httpClient->get(path, request);
+        xml xmlListContainerResponse = <xml>check handleResponse(response);
```
Comment (RolandHewage) [inline modules/blobs/endpoint.bal:59]: xml xmlListContainerResponse = check handleResponse(response);
will also work

File: modules/blobs/endpoint.bal [hunk 8]
```diff
@@ -0,0 +1,939 @@
...
+        http:Request request = createRequest(optionalHeaderMap);
+        map<string> uriParameterMap = {};
+        uriParameterMap[RESTYPE] = ACCOUNT;
+        uriParameterMap[COMP] = PROPERTIES;
+
+        check prepareAuthorizationHeader(request, GET, self.authorizationMethod, self.accountName, self.accessKey, 
+                                            EMPTY_STRING, uriParameterMap);
+        string resourcePath = FORWARD_SLASH_SYMBOL;
+        string path = preparePath(self.authorizationMethod, self.sharedAccessSignature, uriParameterMap, resourcePath);  
+        http:Response response = <http:Response> <http:Response>check self.httpClient->get(path, request);
```
Comment (RolandHewage) [inline modules/blobs/endpoint.bal:168]: <http:Response> is repeated twice

**Review by RolandHewage — COMMENTED** [review]
Few minor issues to be fixed.

File: modules/blobs/data_mappings.bal (hunk 1 above)
Comment (LakshanSS) [reply modules/blobs/data_mappings.bal:60]: Since they are mandatory fields, we won't get that error.

**Comment by anupama-pathirage** [comment]
@LakshanSS The commit messages and PR heading is not according to the best practices. Please correct them.

A properly formed Git commit subject line should always be able to complete the following sentence:

If applied, this commit will **_your subject line here_**

Refer [1] for more details.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md#submit-your-contribution

**Review by kasun04 — APPROVED** [review]

**Pull request merged**